### PR TITLE
openthread: zigbee: lib: Extract ram power down function for common use

### PR DESCRIPTION
**Pull request by MarekPorwisz**
To minimize power consumption thread needs to be able to disable unused
ram regions. After some investigation such function was found in zigbee
helpers.
Moved this function to common libraries and made use of it in thread
coap client sample.

Signed-off-by: Marek Porwisz <marek.porwisz@nordicsemi.no>

### CODE REVIEW

File: lib/ram_pwrdn/ram_pwrdn.c [hunk 1]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
Comment (rlubos) [inline lib/ram_pwrdn/ram_pwrdn.c:2]: 2020 or 2019 - 2020 since it's moved from one place to another.

File: lib/ram_pwrdn/Kconfig [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+config RAM_POWER_DOWN_LIBRARY
+	bool "Enable api for turingn off unused RAM segments"
```
Comment (greg-fer) [inline lib/ram_pwrdn/Kconfig:8]: Typo: `turingn` > `turning`

Comment (greg-fer) [inline lib/ram_pwrdn/Kconfig:8]: `api` or `API`?

File: include/ram_pwrdn.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+/** @file
+ * @brief RAM section power down header.
+ */
+
+#ifndef RAM_PWRDN_H__
+#define RAM_PWRDN_H__
+
+/**
+ * @defgroup ram_pwrdn RAM Power Down
```
Comment (greg-fer) [inline include/ram_pwrdn.h:15]: Spotted in .rst, but I see it's also an issue in other files in this PR: lots of capitalization inconsistencies -- RAM, ram, api, API, ID, id, and so on. Please be consistent.

File: lib/ram_pwrdn/Kconfig (hunk 2 above)
Comment (pszkotak) [reply lib/ram_pwrdn/Kconfig]: Fixed.

File: include/ram_pwrdn.h (hunk 3 above)
Comment (pszkotak) [reply include/ram_pwrdn.h:15]: Fixed.

File: lib/ram_pwrdn/Kconfig (hunk 2 above)
Comment (pszkotak) [reply lib/ram_pwrdn/Kconfig]: API. Fixed.

File: lib/ram_pwrdn/Kconfig [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+config RAM_POWER_DOWN_LIBRARY
+	bool "Enable API for turning off unused RAM segments"
+	help
+	  This allows application to call API for disabling unused RAM segments
```
Comment (greg-fer) [inline lib/ram_pwrdn/Kconfig:10]: Period missing at the end of the line?

File: lib/ram_pwrdn/ram_pwrdn.c [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	uint32_t bank_addr = RAM_START_ADDR + bank_id
+			  * RAM_BANK_0_7_SECTION_SIZE
+			  * RAM_BANK_0_7_SECTIONS_NBR;
+	return bank_addr;
+}
+
+/**@brief Calculate bottom address of section of RAM bank with given bank
+ *        and section IDs.
+ *
+ * @param[in] bank_id     ID of RAM bank sectionn is placed in.
```
Comment (greg-fer) [inline lib/ram_pwrdn/ram_pwrdn.c:55]: Typo: `section[n]`

File: include/ram_pwrdn.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * @{
+ * @brief Module for disabling unused sections of RAM
```
Comment (greg-fer) [inline include/ram_pwrdn.h:17]: Period missing.

File: lib/ram_pwrdn/Kconfig (hunk 4 above)
Comment (pszkotak) [reply lib/ram_pwrdn/Kconfig:10]: Fixed.

File: lib/ram_pwrdn/ram_pwrdn.c (hunk 5 above)
Comment (pszkotak) [reply lib/ram_pwrdn/ram_pwrdn.c:55]: Fixed.

File: include/ram_pwrdn.h (hunk 6 above)
Comment (pszkotak) [reply include/ram_pwrdn.h:17]: Fixed.

**Comment by pszkotak** [comment]
The functionality seems to work just fine. I measured the MTD CoAP Client in SED mode sleep. Sleep current is reduced by ~300nA which is ~7%.

**Comment by sebastiandraus** [comment]
@MarekPorwisz  One comment from my side:
I have asked @ioannisg about using the `_image_ram_end` symbol when I was implementing powering down RAM for Zigbee samples. There might be rare cases where SRAM areas outside zephyr image might be used indirectly by applications but powered down.
Now this implementation is available for other modules/samples so we should add some sort of warning that this situation may happen.

**Comment by pszkotak** [comment]
> here might be rare cases where SRAM areas outside zephyr image might be used indirectly by applications but powered down.
@sebastiandraus, could you mention at least one of such cases so I could include it?

**Comment by sebastiandraus** [comment]
> > here might be rare cases where SRAM areas outside zephyr image might be used indirectly by applications but powered down.
> > @sebastiandraus, could you mention at least one of such cases so I could include it?

@pszkotak When memory is accessed through the pointer, let's assume to address 0x20030000 (outside the `_image_ram_end`). Data can be written and read using the pointer. This ram region will be powered down so the next read/write attempt will result in Bus Fault.
Example log from modified light switch sample:

> E: Writing value: 1
> E: Value read: 1
> E: Writing value: 2
> E: Value read: 2
> E: Powering down the RAM.
> E: Writing value: 3
> E: ***** BUS FAULT *****
> E:   Imprecise data bus error
> E: r0/a1:  0x0002b027  r1/a2:  0x00000045  r2/a3:  0x00000003
> E: r3/a4:  0x20030000 r12/ip:  0x00000000 r14/lr:  0x000008f1
> E:  xpsr:  0x21000000
> E: Faulting instruction address (r15/pc): 0x0002362a
> E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
> E: Current thread: 0x20000560 (unknown)
> E: Resetting system

**Comment by pszkotak** [comment]
> > > here might be rare cases where SRAM areas outside zephyr image might be used indirectly by applications but powered down.
> > > @sebastiandraus, could you mention at least one of such cases so I could include it?
> 
> @pszkotak When memory is accessed through the pointer, let's assume to address 0x20030000 (outside the `_image_ram_end`). Data can be written and read using the pointer. This ram region will be powered down so the next read/write attempt will result in Bus Fault.
> Example log from modified light switch sample:
> 
> > E: Writing value: 1
> > E: Value read: 1
> > E: Writing value: 2
> > E: Value read: 2
> > E: Powering down the RAM.
> > E: Writing value: 3
> > E: ***** BUS FAULT *****
> > E:   Imprecise data bus error
> > E: r0/a1:  0x0002b027  r1/a2:  0x00000045  r2/a3:  0x00000003
> > E: r3/a4:  0x20030000 r12/ip:  0x00000000 r14/lr:  0x000008f1
> > E:  xpsr:  0x21000000
> > E: Faulting instruction address (r15/pc): 0x0002362a
> > E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
> > E: Current thread: 0x20000560 (unknown)
> > E: Resetting system

RST updated.

File: include/ram_pwrdn.rst [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+RAM Power Down
+##############
+
+The RAM Power Down library is a basic module for disabling unused sections of RAM, which allows to save power in low power applications.
+
+To disable unused RAM sections, call :cpp:func:`power_down_unused_ram`.
+This function will automatically disable all memory regions that are not used by the application.
+
+.. note:: If the RAM memory is used indirectly (e.g. through the pointer) it could happen that this region will be powered off by :cpp:func:`power_down_unused_ram`. Next access to that memory region would result in the Bus fault exception.
```
Comment (greg-fer) [inline include/ram_pwrdn.rst:11]: After discussing on Teams:
```suggestion
.. note::
    Make sure you are not writing to memory indirectly, for example through pointers.
    :cpp:func:`power_down_unused_ram` will also power down regions that are using RAM memory indirectly, which can result in a bus fault exception on the next access.
```

Comment (pszkotak) [reply include/ram_pwrdn.rst]: Done.

Comment (sebastiandraus) [reply include/ram_pwrdn.rst]: @greg-fer This is not precise. You can access memory through pointer, e.g when the pointer points to some variable known to linker. If you access memory outside the range linker knows is being used, you may access ram region which is powered off and bus fault will occur.

File: lib/ram_pwrdn/Kconfig [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+	help
+	  This allows application to call API for disabling unused RAM segments.
+	  This is usually not needed, but can improve battery lifetime for
+	  applications that spend most of the time in the sleep mode with most
+	  peripherals disabled.
+
+# Configure RAM_POWERDOWN_LOG_LEVEL
+module =  RAM_POWERDOWN
+module-str = RAM Power down module
+source "${ZEPHYR_BASE}/subsys/logging/Kconfig.template.log_config"
```
Comment (tejlmand) [inline lib/ram_pwrdn/Kconfig:18]: Please use new syntax:
```suggestion
source "$(ZEPHYR_BASE)/subsys/logging/Kconfig.template.log_config"
```

From: https://pypi.org/project/kconfiglib/#library-documentation
```
environment variables can be referenced both as $(FOO) (the new syntax) 
```

Comment (tejlmand) [inline lib/ram_pwrdn/Kconfig:18]: Please place the log level setting under RAM Power down menu, as is done here:
https://github.com/nrfconnect/sdk-nrf/blob/master/lib/dk_buttons_and_leds/Kconfig#L11-L33

File: samples/zigbee/light_switch/src/main.c [hunk 9]
```diff
@@ -429,7 +430,7 @@ void main(void)
 #if defined BUTTON_SLEEPY
 	if (dk_get_buttons() & BUTTON_SLEEPY) {
 		zigbee_configure_sleepy_behavior(true);
-		zigbee_power_down_unused_ram();
+		power_down_unused_ram();
```
Comment (tejlmand) [inline samples/zigbee/light_switch/src/main.c:433]: Should this call be safeguarded with `if (IS_ENABLED(CONFIG_RAM_POWER_DOWN_LIBRARY))` ?

File: lib/ram_pwrdn/Kconfig (hunk 8 above)
Comment (pszkotak) [reply lib/ram_pwrdn/Kconfig]: Fixed.

File: samples/zigbee/light_switch/src/main.c (hunk 9 above)
Comment (pszkotak) [reply samples/zigbee/light_switch/src/main.c]: Fixed.

File: lib/ram_pwrdn/Kconfig (hunk 8 above)
Comment (pszkotak) [reply lib/ram_pwrdn/Kconfig]: Fixed.

Comment (tejlmand) [reply lib/ram_pwrdn/Kconfig]: Not completely, you forgot to change: `config RAM_POWER_DOWN_LIBRARY`
https://github.com/nrfconnect/sdk-nrf/blob/0b69744149ad5eea8fd65b74e312d8afe63fb760/lib/ram_pwrdn/Kconfig#L7

into a menuconfig as here: https://github.com/nrfconnect/sdk-nrf/blob/master/lib/dk_buttons_and_leds/Kconfig#L7

Comment (pszkotak) [reply lib/ram_pwrdn/Kconfig]: I missed that, thanks!
Fixed now.

**Pull request merged**